civet/compilers/base_compiler.py

```diff
@@ -0,0 +1,217 @@
+
+    def on_moved(self, event):
+        if event.is_directory:
+            print(
+                'Warning: Directory %s deleted' % event.src_path,
+                file=sys.stderr)
+        elif (self.compiler.matches(*os.path.splitext(event.src_path))
+              and self.compiler.matches(*os.path.splitext(event.dest_path))):
+            print(
+                'Warning: File renamed {0} -> {1}'.format(
+                    event.src_path, event.dest_path), file=sys.stderr)
+            self.compile(event.dest_path)
+
+
+class Compiler(object):
+    def __init__(self, precompiled_assets_dir, kill_on_error):
+        self.precompiled_assets_dir = precompiled_assets_dir
+        if not hasattr(self, 'executable'):
+            bin = getattr(
+                settings, self.executable_setting, self.executable_name)
+            self.executable = find_executable(bin)
+        if not self.executable:
+            if getattr(settings, self.executable_setting, None):
+                print(
+                    'Your project uses {name}, but "{bin_path}" specified in '
+                    'settings.{settings_name} is not found.'
+                    .format(
+                        name=self.name,
+                        bin_path=bin,
+                        settings_name=self.executable_setting
+                    ), file=sys.stderr)
+            else:
+                print(
+                    'Your project uses {name}, but "{executable_name}" is not '
+                    'found in your PATH.'
+                    .format(
+                        name=self.name,
+                        executable_name=self.executable_name
+                    ), file=sys.stderr)
+            raise_error_or_kill(kill_on_error)
+
+    @property
+    def name(self):
+        """Human visible name for this kind of compiler (eg "CoffeeScript").
+        """
+        raise NotImplementedError("Subclasses must implement name.")
+
+    @property
+    def executable_name(self):
+        """Bare name of compiler executable (eg `coffee`).
+        """
+        raise NotImplementedError("Subclasses must implement executable_name.")
+
+    @property
+    def executable_setting(self):
+        """Django settings name for executable path
+        """
+        raise NotImplementedError(
+            "Subclasses must implement executable_setting.")
+
+    def matches(self, base, ext):
+        """Return true if given base path and file extension is handled by this
+        compiler.
+        """
+        raise NotImplementedError("Subclasses must implement matches()")
+
+    def get_dest_path(self, base, ext):
+        """Return destination path for given filename base and ext (previously
+        accepted by matches().
+        """
+        raise NotImplementedError("Subclasses must implement get_dest_path()")
+
+    def get_arguments(self, src_path, dst_path):
+        """Return compiler executable and arguments needed to compile src_path
+        to dst_path.
```

Comment:
Minor nit: When I read the method name, I thought this would *only* return the arguments to the command, but as the docstring says, this returns both the executable path and the arguments. Maybe this could be renamed to `get_command_with_arguments`?

That said, according to the [Python docs on `subprocess`](https://docs.python.org/2/library/subprocess.html), they call this list of command with arguments just `args`, so maybe this is fine.